text/0003-templates.md

```diff
@@ -0,0 +1,393 @@
+
+### Dependencies
+
+For dependencies specifically, inconsistent versions of the same dependency within a monorepo cause different flavors of problems:
+
+- In projects that bundle dependencies, multiple versions inflate the size of the final result deployed to users.
+- Differing versions result in multiple copies that may not interact well at runtime, especially if features like [`Symbol()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol) are used. For example, [React hooks will error if a component is rendered with a different copy of React in the same app](https://reactjs.org/warnings/invalid-hook-call-warning.html#duplicate-react).
+- For TypeScript, multiple copies of the same `@types` package causes compile errors from mismatching definitions. The compiler diagnostic for this is usually: *"argument of type `Foo` is not assignable to parameter of type `Foo`"*. For developers that have seen this before, they may realize this diagnostic is due to a dependency version mismatch. For developers new to TypeScript, *"`Foo` is not assignable to `Foo`"* is very confusing.
+
+While there are situations differing versions are intentional, this is more often accidental. Multiple differing versions arise from not reviewing `pnpm-lock.yaml` file changes or not searching for existing dependency specifiers before adding a new one. The later is typically unwritten convention in most monorepos.
+
+### Authoring Fields
+
+Fields such as `author`, `license`, and `repository` are typically expected to be the same across all packages in a monorepo. It would be easier to set this in a singular source of truth rather than every package.
+
+### Workspace Configuration
+
+Workspace settings such as [`pnpm.packageExtensions`](https://pnpm.io/package_json#pnpmpackageextensions) can be shared across different repositories. For example, pnpm includes the [`@yarnpkg/extensions` database](https://github.com/yarnpkg/berry/blob/master/packages/yarnpkg-extensions/sources/index.ts) builtin. Templates allow users to create their own extensions database.
+
+## Detailed Explanation
+
+A "*Template*" is a normal `package.json` file with the `pnpm.template` field set.
+
+```json5
+{
+  "name": "@example/frontend-catalog",
+  "pnpm": {
+    "template": "catalog"
+  },
+  "dependencies": {
+    "react": "^18.2.0",
+    "react-dom": "^18.2.0",
+    "redux": "^4.2.0",
+    "react-redux": "^8.0.0"
+  }
+}
+```
+
+A package can then reference the template to populate different portions of its definition. Templates are available in different flavors that provide distinct features.
+
+- `catalog` — templates that share dependency specifiers
```

Comment:
Why `catalog` instead of `dependencies`?

Reply:
Another pnpm user called out a similar feature in Gradle called catalogs. https://github.com/pnpm/pnpm/issues/2713#issuecomment-1341476866

More on the history: https://github.com/pnpm/rfcs/pull/1#issuecomment-1399644823

> For the name, I saw a suggestion for `catalog:` and liked it a lot. [pnpm/pnpm#2713 (comment)](https://github.com/pnpm/pnpm/issues/2713#issuecomment-1341476866). (Or maybe `version-catalog:` so it's more easily searchable online.) The "environment" terminology might overlap with the existing [`pnpm env`](https://pnpm.io/cli/env) command. I think most people will have a good idea what "shareable version catalog package" is.

Credit to @dsilvasc for this.